core/src/test/scala/sbt/librarymanagement/VersionNumberSpec.scala

```diff
@@ -1,150 +1,185 @@
-
-  "0.1.0-RC1" should "be parsed" in beParsedAs("0.1.0-RC1", Seq(0, 1, 0), Seq("RC1"), Seq())
-  it should "cascade" in generateCorrectCascadingNumbers("0.1.0-RC1",
-                                                         Seq("0.1.0-RC1", "0.1.0", "0.1"))
-
-  "0.1.0-MSERVER-1" should "be parsed" in beParsedAs("0.1.0-MSERVER-1",
-                                                     Seq(0, 1, 0),
-                                                     Seq("MSERVER", "1"),
-                                                     Seq())
-  it should "cascade" in generateCorrectCascadingNumbers("0.1.0-MSERVER-1",
-                                                         Seq("0.1.0-MSERVER-1", "0.1.0", "0.1"))
-
-  "2.10.4-20140115-000117-b3a-sources" should "be parsed" in {
-    beParsedAs("2.10.4-20140115-000117-b3a-sources",
-               Seq(2, 10, 4),
-               Seq("20140115", "000117", "b3a", "sources"),
-               Seq())
-  }
-  it should "cascade" in generateCorrectCascadingNumbers(
-    "2.10.4-20140115-000117-b3a-sources",
-    Seq("2.10.4-20140115-000117-b3a-sources", "2.10.4", "2.10"))
-  it should "be SemVer compat with 2.0.0" in beSemVerCompatWith(
-    "2.10.4-20140115-000117-b3a-sources",
-    "2.0.0")
-  it should "be not SecSeg compat with 2.0.0" in notBeSecSegCompatWith(
-    "2.10.4-20140115-000117-b3a-sources",
-    "2.0.0")
-
-  "20140115000117-b3a-sources" should "be parsed" in {
-    beParsedAs("20140115000117-b3a-sources", Seq(20140115000117L), Seq("b3a", "sources"), Seq())
-  }
-  it should "cascade" in generateCorrectCascadingNumbers("20140115000117-b3a-sources",
-                                                         Seq("20140115000117-b3a-sources"))
-
-  "1.0.0-alpha+001+002" should "be parsed" in {
-    beParsedAs("1.0.0-alpha+001+002", Seq(1, 0, 0), Seq("alpha"), Seq("+001", "+002"))
-  }
-  it should "cascade" in generateCorrectCascadingNumbers(
-    "1.0.0-alpha+001+002",
-    Seq("1.0.0-alpha+001+002", "1.0.0", "1.0")
-  )
-
-  "non.space.!?string" should "be parsed" in {
-    beParsedAs("non.space.!?string", Seq(), Seq(), Seq("non.space.!?string"))
-  }
-  it should "cascade" in generateCorrectCascadingNumbers("non.space.!?string",
-                                                         Seq("non.space.!?string"))
-
-  "space !?string" should "be parsed as an error" in beParsedAsError("space !?string")
-  "blank string" should "be parsed as an error" in beParsedAsError("")
-
-  def beParsedAs(s: String, ns: Seq[Long], ts: Seq[String], es: Seq[String]) =
-    s match {
-      case VersionNumber(ns1, ts1, es1) if (ns1 == ns && ts1 == ts && es1 == es) =>
-        (VersionNumber(ns, ts, es).toString shouldBe s)
-        (VersionNumber(ns, ts, es) shouldBe VersionNumber(ns, ts, es))
+// This is a specification to check VersionNumber and VersionNumberCompatibility.
+class VersionNumberSpec extends UnitSpec with Inside {
+  import VersionNumber.{ SemVer, SecondSegment }
+
+  version("1") { implicit v =>
```

Comment:
The overall looks is cleaner, but I am not really sure if I like `implicit` used just so it doesn't pass in one parameter.

Reply:
I can change that implementation detail

Reply:
Another sort of meta question is: Do we make custom DSL for each spec / unit tests we write that reads like English statements? I'd much rather prefer we used something like JUnit or Minitest, if we are moving away from plain UnitSpec.

Reply:
I'm happy to discuss and/or experiment with alternatives to UnitSpec.

but note that in this PR:
but I'm not moving away from UnitSpec in this PR, I'm just removing extreme repetition

Reply:
I'm happy to discuss and/or experiment with alternatives to UnitSpec.

but note that I'm not moving away from UnitSpec in this PR.

I am in favour of creating helper methods to remove lots of repetition and improve legibility (and maintainability), but the fact that it reads like English statements is not my objective.

I'm not sure where the line is drawn between creating a few helper methods and removing repetition becomes a "custom DSL".

Reply:
```scala
"1" should "be parsed" in beParsedAs("1", Seq(1), Seq(), Seq())
it should "breakdown" in breakDownTo("1", Some(1))
```

In the above, `beParsedAs` is an English style helper function idiomatic to ScalaTest's `FlatSpec`.

> but note that I'm not moving away from UnitSpec in this PR.

I'd say you've made our own SomethingSpec that internally uses `FlatSpec`. I really like it, but I also want to stick to an existing test framework, esp if we want others to follow it. Perhaps the style we should migrate is `FunSuite`:

```scala
  test("1") {
    val v = "1"
    assertParsesTo(v, Seq(1), Seq(), Seq())
    assertBreaksDownTo(v, Some(1))
    assertCascadesTo(v, Seq("1"))
  }
```

Less noise. WDYT?